text/0813-template-layering.md

```diff
@@ -0,0 +1,1001 @@
+Relevant Team(s): Ember.js
+RFC PR: https://github.com/emberjs/rfcs/pull/813
+---
+
+# Layering and Desugaring for First-Class Component Templates
+
+## Summary
+
+This RFC propose to introduce a `template()` function as an substitution for
+the `<template>` syntax extension defined in [RFC #779][rfc-779] using
+standard JavaScript, for use in environments where the custom syntax extension
+is not available, and also as a publishing format for addons on NPM:
+
+[rfc-779]: ./0779-first-class-component-templates.md
+
+```js
+import { template } from "@ember/template-compilation";
+import Hello from "my-app/components/hello";
+
+// <template><Hello /></template> becomes...
+export default template(`<Hello />`, () => ({ Hello }));
+
+// export const Foo = <template><Hello /></template> becomes...
+export const Foo = template(`<Hello />`, () => ({ Hello }));
+
+// export class Bar {
+//   <template><Hello /></template>
+// }
+export class Bar {
+  static {
+    template(`<Hello />`, () => ({ Hello }), this);
+  }
+}
+```
+
+Idiomatic usages of the `template()` function can be pre-compiled by a build
+step if desired, but a runtime implementation can also be used via an opt-in
+described below.
+
+We also propose an alternative form that uses the `template()` function as a
+tag function for JavaScript tagged template strings. However, this variant does
+not have a runtime implementation and must be handled by a build step:
+
+```js
+import { template } from "@ember/template-compilation";
+import Hello from "my-app/components/hello";
+
+// <template><Hello /></template> becomes...
+export default template`<Hello />`;
+
+// export const Foo = <template><Hello /></template> becomes...
+export const Foo = template`<Hello />`;
+
+// export class Bar {
+//   <template><Hello /></template>
+// }
+export class Bar {
+  static {
+    template`<Hello />`;
+  }
+}
+```
+
+In addition, we also propose a new `@ember/template-compilation/runtime`
+module as an opt-in to enable runtime template compilation.
+
+## Motivation
+
+[RFC #779][rfc-779] introduced a first-class component syntax feature that has
+numerous benefits, among which are:
+
+1. Opt-in to strict mode ([RFC #496][rfc-496])
+2. Eliminating the need for name-based runtime resolutions
+3. Ability to use imported template constructs
+4. Ability to interact with the surrounding JavaScript scope
+
+[rfc-496]: ./0496-handlebars-strict-mode.md
+
+However, one of the drawbacks is that it requires a custom syntax extension
+(`<template>`) and a custom file format (`.gjs/.gts`), which requires a build
+step and other custom tooling.
+
+RFC #779 makes a good case for why this is necessary and preferable to the
+alternatives. We stand by the rationales given in RFC #779, and this RFC does
+not intend to revisit its recommendations. However, there remain cases where
+this drawback is highly undesirable or simply not acceptable:
+
+1. Until we have implemented good editor integration, the developer experience
```

Comment:
[question] I thought we accepted that editor integration (to include syntax highlighting etc) was a blocker to implementation (with the idea that it would drive implementation). When did that change?